File: Data Structures & Algorithms/max-water-container/submission-0.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maxArea(vector<int>& heights) {
        int l=0, r= heights.size()-1;
        int m = min(heights[l],heights[r])*(r-l);

        while (l<r)
        {
            int temp = vol(heights, l, r);
            m = max(m, temp);

            if(heights[l]<=heights[r])
                l++;
            
            else
                r--;
            
        }
        return m;
    }

    int vol(vector<int>& heights, int l, int r)
    {
        return min(heights[l],heights[r])*(r-l);
    }
};

File: Data Structures & Algorithms/trapping-rain-water/submission-0.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0

        l,r = 0, len(height)-1
        lm,rm = height[l],height[r]

        ans=0

        while l<r:
            if lm<rm:
                l+=1
                lm=max(lm,height[l])
                ans+=lm-height[l]
            else:
                r-=1
                rm=max(rm,height[r])
                ans+=rm-height[r]
        return ans